doc/source/serve/performance.rst

```diff
@@ -0,0 +1,81 @@
+  to learn more about the maximum number of nodes and other limitations.
+
+You can check out our `microbenchmark instruction <https://github.com/ray-project/ray/blob/master/python/ray/serve/benchmarks/README.md>`_
+to benchmark on your hardware.
+
+Debug performance Issue
+-----------------------
+A typical performance issue in Serve has the symptom of high client call latency and low throughput.
+This shows the system is processing fewer queries than desired! 
+
+If you have set up :ref:`monitoring <serve-monitoring>` with Ray and Ray Serve, you will observe that
+``serve_num_router_requests`` is constant while your load increases
+``serve_backend_queuing_latency_ms`` is spiking up as queries queue up in the background
+
+Given the symptom, there are several ways to fix it.
+
+Choosing the right hardware
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Make sure you are using the right hardware and resources. 
+Are you using GPUs (``actor_init_options={“num_gpus”: 1}``) or 1+ cores (``actor_init_options={“num_cpus”: 2}``, and setting ``OMP_NUM_THREADS``)
+to increase the performance of your deep learning framework?
+
+Batching
+^^^^^^^^
+If your backend can process a batch at a time at a sublinear latency 
+(for example, if it takes 1ms to process 1 query and 5ms to process 10 of them) 
+then batching is your best approach. Check out the :ref:`batching guide <serve-ml-batching>` to 
+make your backend accept batches. You might want to tune your ``max_batch_size`` and ``batch_wait_timeout`` to maximize the benefits:
+
+- ``max_batch_size`` specifies how big the batch should be. Generally, 
+   we recommend choosing the largest batch size your function can handle 
+   AND the performance improvement is no longer sublinear. Take a dummy 
+   example: suppose it takes 1ms to process 1 query, 5ms to process 10 queries,
+   and 6ms to process 11 queries. Here you should set the batch size to to 10 
+   because adding more queries won’t improve the performance.
+- ``batch_wait_timeout`` specifies how the maximum amount of time to wait before
+  a batch should be processed, even if it’s not full.  It should be set according 
+  to `batch-wait-timeout + full batch processing time ~= expected latency`. The idea 
+  here is to have the first query wait for the longest possible time to achieve high throughput.  
+  This means you should set batch-wait-timeout as large as possible without exceeding your desired expected latency in the equation above.
+
+Async functions
+^^^^^^^^^^^^^^^
+Are you using ``async def`` in your callable? If you are using asyncio and
+hitting the same queuing issue mentioned above, you might want to increase 
+``max_concurrent_queries``. Serve sets a low number by default so the client gets 
+proper backpressure. You can increase the value in the :mod:`backend config <ray.serve.config.BackendConfig>`
+to allow more coroutines running in the same replica.
+
+Scale HTTP servers
```

Comment:
Do we have any good way to diagnose that this is the issue? Can we export some metrics to make it clear?

Reply:
hmm probably just the CPU utilization for the HTTP servers?